src/quicklist.c

```diff
@@ -473,23 +534,45 @@ REDIS_STATIC int _quicklistNodeAllowMerge(const quicklistNode *a,
 
 #define quicklistNodeUpdateSz(node)                                            \
     do {                                                                       \
-        (node)->sz = ziplistBlobLen((node)->zl);                               \
+        (node)->sz = ziplistBlobLen((node)->entry);                               \
     } while (0)
 
+static quicklistNode* __quicklistCreatePlainNode(void *value, size_t sz) {
+    quicklistNode *new_node = quicklistCreateNode();
+    new_node->entry = zmalloc(sz);
+    new_node->container = QUICKLIST_NODE_CONTAINER_NONE;
+    memcpy(new_node->entry , value, sz);
+    new_node->sz = sz;
+    new_node->count++;
+    return new_node;
+}
+
+static void __quicklistInsertPlainNode(quicklist *quicklist, quicklistNode *old_node,
+                                       void *value, size_t sz, int after)  {
+    __quicklistInsertNode(quicklist, old_node, __quicklistCreatePlainNode(value, sz), after);
+    quicklist->count++;
+}
+
 /* Add new entry to head node of quicklist.
  *
  * Returns 0 if used existing head.
  * Returns 1 if new head created. */
 int quicklistPushHead(quicklist *quicklist, void *value, size_t sz) {
     quicklistNode *orig_head = quicklist->head;
+
+    if(unlikely(isLargeElement(sz))) {
```

Comment:
If `quicklist->head` is `QUICKLIST_NODE_CONTAINER_NONE`, you also need to insert the plain node directly.
The same applies to `quicklistPushTail`.

Reply:
not sure the scenario you are referring to... if 
(new node == plain) && (head == plain || head ==small) we get to first if
(new node == small) && (head == plain) second if fails and we get to the else
(new node == small) && (head == small) we will probably get to if (or else if ziplist is full)

Reply:
Yeah, I misunderstood.